libraries/chain/controller.cpp

```diff
@@ -414,180 +938,192 @@ struct controller_impl {
       wlog( "Initializing new blockchain with genesis state" );
-      producer_authority_schedule initial_schedule = { 0, { producer_authority{config::system_account_name, block_signing_authority_v0{ 1, {{genesis.initial_key, 1}} } } } };
-      legacy::producer_schedule_type initial_legacy_schedule{ 0, {{config::system_account_name, genesis.initial_key}} };
-
-      block_header_state_legacy genheader;
-      genheader.active_schedule                = initial_schedule;
-      genheader.pending_schedule.schedule      = initial_schedule;
-      // NOTE: if wtmsig block signatures are enabled at genesis time this should be the hash of a producer authority schedule
-      genheader.pending_schedule.schedule_hash = fc::sha256::hash(initial_legacy_schedule);
-      genheader.header.timestamp               = genesis.initial_timestamp;
-      genheader.header.action_mroot            = genesis.compute_chain_id();
-      genheader.id                             = genheader.header.calculate_id();
-      genheader.block_num                      = genheader.header.block_num();
-
-      head = std::make_shared<block_state_legacy>();
-      static_cast<block_header_state_legacy&>(*head) = genheader;
-      head->activated_protocol_features = std::make_shared<protocol_feature_activation_set>();
-      head->block = std::make_shared<signed_block>(genheader.header);
-      db.set_revision( head->block_num );
+
+      auto init_blockchain = [&genesis](auto& fork_db, auto& head) {
+         producer_authority_schedule initial_schedule = { 0, { producer_authority{config::system_account_name, block_signing_authority_v0{ 1, {{genesis.initial_key, 1}} } } } };
+         legacy::producer_schedule_type initial_legacy_schedule{ 0, {{config::system_account_name, genesis.initial_key}} };
+
+         block_header_state_legacy genheader;
+         genheader.active_schedule                = initial_schedule;
+         genheader.pending_schedule.schedule      = initial_schedule;
+         // NOTE: if wtmsig block signatures are enabled at genesis time this should be the hash of a producer authority schedule
+         genheader.pending_schedule.schedule_hash = fc::sha256::hash(initial_legacy_schedule);
+         genheader.header.timestamp               = genesis.initial_timestamp;
+         genheader.header.action_mroot            = genesis.compute_chain_id();
+         genheader.id                             = genheader.header.calculate_id();
+         genheader.block_num                      = genheader.header.block_num();
+
+         head = std::make_shared<block_state_legacy>();
+         static_cast<block_header_state_legacy&>(*head) = genheader;
+         head->activated_protocol_features = std::make_shared<protocol_feature_activation_set>();
+         head->block = std::make_shared<signed_block>(genheader.header);
+      };
+
+      block_data.apply_dpos<void>(init_blockchain); // [greg todo] assuming here that genesis_state is always dpos
```

Comment:
This todo can be removed. We have no immediate goal of supporting IF at genesis.

Reply:
Remeved the `[todo]` part.